docs/proposals/ssh_adapter.md

```diff
@@ -0,0 +1,194 @@
+the operation.  This would be inefficient in some server-side implementations,
+and nearly impossible to implement securely for implementations that use the
+system OpenSSH for implementation.
+
+## A more usable approach
+
+Git already has some places we can look for inspiration.  Its SSH protocol is
+based on the Git native protocol, which is based on the pkt-line scheme.
+Recently, Git has learned about protocol version 2, which provides better
+support for expressing and negotiating capabilities.
+
+Ideally, we would allow multiple operations to occur on a single connection for
+efficiency's sake, especially on high-latency connections, where the cost of SSH
+connection setup may be high due to multiple round trips.  In addition, a
+protocol which maps well onto HTTP may be beneficial for those server-side
+implementations which would like to proxy connections to an HTTP-based backend.
+
+This design assumes a reference to Git's pkt-line and protocol v2 documentation.
+pkt-line headers for this document may contain values up to 65519 decimal.
+
+To initiate a connection, Git LFS should run the folllowing command:
+
+    $ ssh [{user}@]{server} git-lfs-transfer {path} {operation}
+
+If authentication fails, or some other connection error occurs, errors will be
+read fronm standard error and displayed to the user.  The operation may be
+`upload`, `download`, or `meta` (for items which are not transfers).
+
+Once the connection is established, the server should send a capability
+advertisement:
+
+```
+capability-advertisement = capability-list flush-pkt
+capability-list = *capability
+capability = PKT-LINE(key=[value] LF)
+
+key = 1*(ALPHA | DIGIT | "-_")
+value = 1*(ALPHA | DIGIT | " -_.,?\/{}[]()<>!@#$%^&*+=:;")
+```
+
+Unlike the Git protocol, but like IMAP, the protocol version is specified as a
+capability.  This document defines protocol version 1, which is specified as
+`version=1`.  If the server supports other protocol versions, it may enumerate
+them here as well.
+
+The client will then issue an appropriate version command:
+
+```
+version-request = PKT-LINE("version " number LF) flush-pkt
+number = 1*DIGIT
+```
+
+No reply is expected.
+
+### Requests to transfer objects
+
+These commands may be used if the operation was `upload` or `download`.
+
+The `batch` command is used to specify a JSON command identical those used at
+the `info/lfs/object/batch` endpoint:
+
+```
+batch-request = batch-command
+                *arguments
+                delim-pkt
+                json-message
+                flush-pkt
+batch-command = PKT-LINE("batch" LF)
+arguments = PKT-LINE(key=[value] LF)
+json-message = 1*PKT-LINE(json-data)
+```
+
+It is recommended to specify a `size` argument, which is the number of bytes in
+decimal of the JSON payload without pkt-line encoding.  A server may signal an
+error if it requires a `size` argument for an operation but one is not specified.
+
+Unknown arguments should be ignored.
```

Comment:
I think that it would be OK to terminate the connection here, depending on your thoughts, too. Unknown arguments mean (to me) undefined behavior, which is not something that we should be subjecting the user to.

Reply:
I think allowing additional arguments that are unknown allows us to do things like adding support for `refname` without needing to bump the protocol.  If the behavior is not backwards compatible, then we'd need a new protocol version or a new capability.

Reply:
Ah, that makes sense. I think this is OK as-is, then.